core/src/main/java/com/linecorp/armeria/internal/server/annotation/FileAggregatedMultipart.java

```diff
@@ -48,43 +51,45 @@ ListMultimap<String, String> params() {
         return params;
     }
 
-    ListMultimap<String, Path> files() {
+    ListMultimap<String, MultipartFile> files() {
         return files;
     }
 
     static CompletableFuture<FileAggregatedMultipart> aggregateMultipart(ServiceRequestContext ctx,
                                                                          HttpRequest req) {
-        final Path multipartUploadsLocation = ctx.config().multipartUploadsLocation();
+        final Path destination = ctx.config().multipartUploadsLocation();
         return Multipart.from(req).collect(bodyPart -> {
-            if (bodyPart.filename() != null) {
-                final ScheduledExecutorService blockingExecutorService =
-                        ctx.blockingTaskExecutor().withoutContext();
-                return resolveTmpFile(multipartUploadsLocation.resolve("incomplete"),
-                                      blockingExecutorService)
-                        .thenComposeAsync(
-                                path -> bodyPart
-                                        .writeTo(path)
-                                        .thenCompose(ignore -> moveFile(
-                                                path,
-                                                multipartUploadsLocation.resolve("complete"),
-                                                blockingExecutorService))
-                                        .thenApply(completePath -> Maps.<String, Object>immutableEntry(
-                                                bodyPart.name(), completePath)),
-                                ctx.eventLoop());
+            final String name = bodyPart.name();
+            final String filename = bodyPart.filename();
+            final EventLoop eventLoop = ctx.eventLoop();
+
+            if (filename != null) {
+                final Path incompleteDir = destination.resolve("incomplete");
+                final ScheduledExecutorService executor = ctx.blockingTaskExecutor().withoutContext();
+
+                return resolveTmpFile(incompleteDir, filename, executor).thenComposeAsync(path -> {
+                    return bodyPart.writeTo(path).thenCompose(ignore -> {
+                        final Path completeDir = destination.resolve("complete");
+                        return moveFile(path, completeDir, executor);
+                    }).thenApply(completePath -> MultipartFile.of(name, filename, completePath.toFile()));
+                }, eventLoop);
```

Comment:
This should be `executor` because it writes to the path? `bodyPart.writeTo(path)`

Reply:
I guess the `eventLoop` seems to be used to get a blocking task executor from the current request context.
The code could be changed to use `thenCompose()` and explicitly pass `executor`, `eventLoop` to `writeTo()`
that will remove an additional context-switching.